module/VuFind/src/VuFind/Db/Service/RatingsService.php

```diff
@@ -40,6 +42,156 @@
  */
 class RatingsService extends AbstractService
 {
+    /**
+     * User database service
+     *
+     * @var \VuFind\Db\Service\UserService
+     */
+    protected $userService;
+
+    /**
+     * Resource database service
+     *
+     * @var \VuFind\Db\Service\ResourceService
+     */
+    protected $resourceService;
+
+    /**
+     * Constructor
+     *
+     * @param EntityManager       $entityManager       Doctrine ORM entity manager
+     * @param EntityPluginManager $entityPluginManager VuFind entity plugin manager
+     * @param ResourceService     $rs                  Database resource service
+     * @param UserService         $us                  Database user service
+     */
+    public function __construct(
+        EntityManager $entityManager,
+        EntityPluginManager $entityPluginManager,
+        \VuFind\Db\Service\ResourceService $rs,
+        \VuFind\Db\Service\UserService $us
+    ) {
+        parent::__construct($entityManager, $entityPluginManager);
+        $this->resourceService = $rs;
```

Comment:
It might be worth considering passing in the service plugin manager, rather than the individual services. I'm not saying that this is something you MUST change, just something to think about.

Advantages to the current approach:
  - Very explicit, makes the dependency relationships very clear

Advantages to passing the service manager:
  - More performant (we only create services on demand when we need them, rather than every time we interact with the ratings service)
  - More flexible (if we need to access additional services in the future, we don't need to change the constructor and factory to allow it)

If you do think it is worth changing, you could also consider refactoring your changes to the RowGateway class into a \VuFind\Db\Service\ServiceAwareTrait, to ensure that we interact with the service plugin manager consistently across classes. (You could still make the service plugin manager a constructor parameter here, but then you could call the trait's setter from this class's constructor).

Reply:
Update: if you take my suggestion below about creating a ServiceAwareInterface and auto-injecting the service plugin manager in the initializer, then you can potentially eliminate the need for this whole factory, since implementing the ServiceAwareInterface and associated trait in the Ratings class will cause the plugin manager to be auto-injected. I think that's another argument toward taking that approach, since it results in even less code to maintain.